src/Features/Core/Portable/Workspace/CompileTimeSolutionProvider.cs

```diff
@@ -0,0 +1,112 @@
+
+        private Solution? _lazyCompileTimeSolution;
+        private int _correspondingDesignTimeSolutionVersion;
+        private readonly bool _enabled;
+
+        public CompileTimeSolutionProvider(Workspace workspace)
+        {
+            _workspace = workspace;
+            _correspondingDesignTimeSolutionVersion = -1;
+            _enabled = workspace.Services.GetRequiredService<IExperimentationService>().IsExperimentEnabled(WellKnownExperimentNames.RazorLspEditorFeatureFlag);
+        }
+
+        private static bool IsRazorAnalyzerConfig(TextDocumentState documentState)
+            => documentState.FilePath != null && documentState.FilePath.EndsWith(RazorEncConfigFileName, StringComparison.OrdinalIgnoreCase);
+
+        public Solution GetCurrentCompileTimeSolution()
+        {
+            if (!_enabled)
+            {
+                return _workspace.CurrentSolution;
+            }
+
+            lock (_guard)
+            {
+                var currentDesignTimeSolution = _workspace.CurrentSolution;
+
+                // Design time solution hasn't changed since we calculated the last compile-time solution:
+                if (currentDesignTimeSolution.WorkspaceVersion == _correspondingDesignTimeSolutionVersion)
+                {
+                    Contract.ThrowIfNull(_lazyCompileTimeSolution);
+                    return _lazyCompileTimeSolution;
+                }
+
+                using var _1 = ArrayBuilder<DocumentId>.GetInstance(out var configIdsToRemove);
+                using var _2 = ArrayBuilder<DocumentId>.GetInstance(out var documentIdsToRemove);
+
+                foreach (var (_, projectState) in currentDesignTimeSolution.State.ProjectStates)
+                {
+                    var anyConfigs = false;
+
+                    foreach (var configState in projectState.AnalyzerConfigDocumentStates.States)
+                    {
+                        if (IsRazorAnalyzerConfig(configState))
+                        {
+                            configIdsToRemove.Add(configState.Id);
+                            anyConfigs = true;
+                        }
+                    }
+
+                    // only remove design-time only documents when source-generated ones replace them
+                    if (anyConfigs)
+                    {
+                        foreach (var documentState in projectState.DocumentStates.States)
+                        {
+                            if (documentState.Attributes.DesignTimeOnly)
+                            {
+                                documentIdsToRemove.Add(documentState.Id);
+                            }
+                        }
+                    }
+                }
+
+                _lazyCompileTimeSolution = currentDesignTimeSolution
+                    .RemoveAnalyzerConfigDocuments(configIdsToRemove.ToImmutable())
+                    .RemoveDocuments(documentIdsToRemove.ToImmutable());
```

Comment:
Note that trying to do a "single" remove operation here rather than just putting it inside the loop over each project state might actually be slower: RemoveDocuments here only works by doing a GroupBy(d => d.ProjectId) and then processing each group individually.

Reply:
OK, I'll revisit.

Reply:
Observations: 
1) The equivalent Remove methods on Project call the Solution ones - so removing multiple documents from each project is worse than from the Solution.
2) Removing each document individually seems to create a lot throw away objects.

Reply:
I'll follow up on this in a separate PR if we conclude there is a better way.